crates/ruff_db/src/file_system.rs

```diff
@@ -0,0 +1,265 @@
+    }
+}
+
+impl AsRef<FileSystemPath> for FileSystemPathBuf {
+    fn as_ref(&self) -> &FileSystemPath {
+        self.as_path()
+    }
+}
+
+impl AsRef<FileSystemPath> for FileSystemPath {
+    #[inline]
+    fn as_ref(&self) -> &FileSystemPath {
+        self
+    }
+}
+
+impl AsRef<FileSystemPath> for str {
+    #[inline]
+    fn as_ref(&self) -> &FileSystemPath {
+        FileSystemPath::new(self)
+    }
+}
+
+impl AsRef<FileSystemPath> for String {
+    #[inline]
+    fn as_ref(&self) -> &FileSystemPath {
+        FileSystemPath::new(self)
+    }
+}
+
+impl AsRef<Path> for FileSystemPath {
+    #[inline]
+    fn as_ref(&self) -> &Path {
+        self.0.as_std_path()
+    }
+}
+
+impl Deref for FileSystemPathBuf {
+    type Target = FileSystemPath;
+
+    fn deref(&self) -> &Self::Target {
+        self.as_path()
+    }
+}
+
+impl std::fmt::Debug for FileSystemPath {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl std::fmt::Display for FileSystemPath {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl std::fmt::Debug for FileSystemPathBuf {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl std::fmt::Display for FileSystemPathBuf {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Metadata {
+    revision: FileRevision,
+    permissions: Option<u32>,
+    file_type: FileType,
+}
+
+impl Metadata {
+    pub fn revision(&self) -> FileRevision {
+        self.revision
+    }
+
+    pub fn permissions(&self) -> Option<u32> {
+        self.permissions
+    }
+
+    pub fn file_type(&self) -> FileType {
+        self.file_type
+    }
+}
+
+/// A number representing the revision of a file.
+///
+/// Two revisions that don't compare equal signify that the file has been modified.
+/// Revisions aren't guaranteed to be monotonically increasing or in any specific order.
+///
+/// Possible revisions are:
+/// * The last modification time of the file.
+/// * The hash of the file's content.
+/// * The revision as it comes from an external system, for example the LSP.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct FileRevision(u128);
+
+impl FileRevision {
+    pub fn new(value: u128) -> Self {
+        Self(value)
+    }
+
+    pub const fn zero() -> Self {
+        Self(0)
+    }
+
+    #[must_use]
+    pub fn as_u128(self) -> u128 {
+        self.0
+    }
+}
+
+impl From<u128> for FileRevision {
+    fn from(value: u128) -> Self {
+        FileRevision(value)
+    }
+}
+
+impl From<u64> for FileRevision {
+    fn from(value: u64) -> Self {
+        FileRevision(u128::from(value))
+    }
+}
+
+impl From<FileTime> for FileRevision {
+    fn from(value: FileTime) -> Self {
+        let seconds = value.seconds() as u128;
```

Comment:
Good catch. I should probably use `from` here too. 

> When converting between numeric types, one thing to note is that From is only implemented for lossless conversions (e.g. you can convert from i32 to i64 with From, but not the other way around), whereas as works for both lossless and lossy conversions (if the conversion is lossy, it truncates). Thus, if you want to ensure that you don't accidentally perform a lossy conversion, you may prefer using From::from rather than as. [source](https://stackoverflow.com/questions/48795329/what-is-the-difference-between-fromfrom-and-as-in-rust)

Reply:
Actually, I have to use `as` for seconds because seconds is a `i64` and converting it to an `u128` doesn't suffer truncation, but the conversion isn't lossless.